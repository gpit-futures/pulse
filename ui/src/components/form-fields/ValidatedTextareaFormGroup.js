import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export default class ValidatedTextareaFormGroup extends PureComponent {
  static propTypes = {
    label: PropTypes.string.isRequired,
    input: PropTypes.object.isRequired,
    meta: PropTypes.shape({
      active: PropTypes.bool,
      error: PropTypes.any,
    }).isRequired,
  };

  state = {
    isChanged: false,
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.meta.dirty) {
      this.setState({ isChanged: true });
    }
  }

  render() {
    const { label, input, meta: { error, touched }, id, isSubmit, isAdvancedSearch } = this.props;
    const { isChanged } = this.state;
    const showError = ((touched || isChanged || isSubmit) && error);

    return (
      <div className={classNames('form-group', { 'has-error': showError }, { 'has-success': isChanged && !error })}>
        <label htmlFor={id} className="control-label">{label}</label>
        <div className="input-holder">
          <textarea
            className="form-control textarea-big input-sm"
            id={id}
            {...input}
          />
        </div>
        {(showError && isAdvancedSearch) ? <span className="required-label">{error}</span> : null}
        {(showError && !isAdvancedSearch) ? <span className="help-block animate-fade">{error}</span> : null}
      </div>
    )
  }
}
