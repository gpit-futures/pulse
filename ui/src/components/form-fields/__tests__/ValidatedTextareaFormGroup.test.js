import React from 'react';
import { configure, shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-15';

import ValidatedTextareaFormGroup from '../ValidatedTextareaFormGroup';

configure({ adapter: new Adapter() });

const testProps = {
  label: 'Test label',
  input: { name: 'test-name', value: '' },
  meta: {
    active: false,
    error: false,
  },
  id: 'test-id',
};

describe('Component <ValidatedTextareaFormGroup />', () => {
  it('should renders with all props correctly', () => {
    const component = shallow(
      <ValidatedTextareaFormGroup
        id={testProps.id}
        label={testProps.label}
        input={testProps.input}
        meta={testProps.meta}
        isSubmit
        isAdvancedSearch={false}
      />);

    expect(component.find('input')).toHaveLength(0);
    expect(component.find('textarea')).toHaveLength(1);
    expect(component.find('label')).toHaveLength(1);

    expect(component.find('label.control-label').text()).toEqual(testProps.label);
    expect(component.find('label.control-label').prop('htmlFor')).toEqual(testProps.id);
    expect(component.find('textarea').prop('id')).toEqual(testProps.id);

    expect(component.instance().props.id).toEqual(testProps.id);
    expect(component.instance().props.label).toEqual(testProps.label);
    expect(component.instance().props.input).toEqual(testProps.input);
    expect(component.instance().props.meta).toEqual(testProps.meta);
    expect(component.instance().props.isSubmit).toEqual(true);
    expect(component.instance().props.isAdvancedSearch).toEqual(false);

    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it('should renders with different state of validation', () => {
    let tree;
    const component = shallow(
      <ValidatedTextareaFormGroup
        id={testProps.id}
        label={testProps.label}
        input={testProps.input}
        meta={testProps.meta}
        isSubmit
        isAdvancedSearch={false}
      />);

    expect(component.find('.form-group').at(0).hasClass('has-error')).toEqual(false);
    expect(component.find('.form-group').at(0).hasClass('has-success')).toEqual(false);
    expect(component.state('isChanged')).toEqual(false);
    tree = toJson(component);
    expect(tree).toMatchSnapshot();

    // Is there a check has-error class
    // 1 case
    component.setProps({
      meta: {
        touched: false,
        error: 'Error',
      },
      isSubmit: true,
    });
    expect(component.find('.form-group').at(0).hasClass('has-error')).toEqual(true);
    expect(component.find('.form-group').at(0).hasClass('has-success')).toEqual(false);
    expect(component.state('isChanged')).toEqual(false);
    tree = toJson(component);
    expect(tree).toMatchSnapshot();

    // 2 case
    component.setProps({
      meta: {
        touched: true,
        error: 'Error',
      },
      isSubmit: false,
    });
    expect(component.find('.form-group').at(0).hasClass('has-error')).toEqual(true);
    expect(component.find('.form-group').at(0).hasClass('has-success')).toEqual(false);
    expect(component.state('isChanged')).toEqual(false);
    tree = toJson(component);
    expect(tree).toMatchSnapshot();

    // 3 case
    component.setProps({
      meta: {
        dirty: true,
        touched: false,
        error: 'Error',
      },
      isSubmit: false,
    });
    expect(component.find('.form-group').at(0).hasClass('has-error')).toEqual(true);
    expect(component.find('.form-group').at(0).hasClass('has-success')).toEqual(false);
    expect(component.state('isChanged')).toEqual(true);
    tree = toJson(component);
    expect(tree).toMatchSnapshot();


    // Is there a check has-success class
    component.setProps({
      meta: {
        dirty: true,
        touched: true,
        error: false,
      },
      isSubmit: true,
    });
    expect(component.find('.form-group').at(0).hasClass('has-error')).toEqual(false);
    expect(component.find('.form-group').at(0).hasClass('has-success')).toEqual(true);
    expect(component.state('isChanged')).toEqual(true);
    tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it('should renders with error message correctly', () => {
    let tree;
    const component = shallow(
      <ValidatedTextareaFormGroup
        id={testProps.id}
        label={testProps.label}
        input={testProps.input}
        meta={{
          dirty: true,
          touched: true,
          error: 'Error',
        }}
        isSubmit
      />);

    expect(component.find('.form-group').at(0).hasClass('has-error')).toEqual(true);

    // 1 case when isAdvancedSearch === false
    expect(component.find('.required-label')).toHaveLength(0);
    expect(component.find('.help-block')).toHaveLength(1);
    expect(component.instance().props.isAdvancedSearch).toEqual(undefined);

    tree = toJson(component);
    expect(tree).toMatchSnapshot();

    // 1 case when isAdvancedSearch === false
    component.setProps({ isAdvancedSearch: true });

    expect(component.find('.required-label')).toHaveLength(1);
    expect(component.find('.help-block')).toHaveLength(0);
    expect(component.instance().props.isAdvancedSearch).toEqual(true);

    tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
