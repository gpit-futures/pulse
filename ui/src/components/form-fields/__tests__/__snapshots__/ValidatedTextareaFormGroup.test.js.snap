// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <ValidatedTextareaFormGroup /> should renders with all props correctly 1`] = `
<div
  className="form-group"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
</div>
`;

exports[`Component <ValidatedTextareaFormGroup /> should renders with different state of validation 1`] = `
<div
  className="form-group"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
</div>
`;

exports[`Component <ValidatedTextareaFormGroup /> should renders with different state of validation 2`] = `
<div
  className="form-group has-error"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
  <span
    className="help-block animate-fade"
  >
    Error
  </span>
</div>
`;

exports[`Component <ValidatedTextareaFormGroup /> should renders with different state of validation 3`] = `
<div
  className="form-group has-error"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
  <span
    className="help-block animate-fade"
  >
    Error
  </span>
</div>
`;

exports[`Component <ValidatedTextareaFormGroup /> should renders with different state of validation 4`] = `
<div
  className="form-group has-error"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
  <span
    className="help-block animate-fade"
  >
    Error
  </span>
</div>
`;

exports[`Component <ValidatedTextareaFormGroup /> should renders with different state of validation 5`] = `
<div
  className="form-group has-success"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
</div>
`;

exports[`Component <ValidatedTextareaFormGroup /> should renders with error message correctly 1`] = `
<div
  className="form-group has-error"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
  <span
    className="help-block animate-fade"
  >
    Error
  </span>
</div>
`;

exports[`Component <ValidatedTextareaFormGroup /> should renders with error message correctly 2`] = `
<div
  className="form-group has-error"
>
  <label
    className="control-label"
    htmlFor="test-id"
  >
    Test label
  </label>
  <div
    className="input-holder"
  >
    <textarea
      className="form-control textarea-big input-sm"
      id="test-id"
      name="test-name"
      value=""
    />
  </div>
  <span
    className="required-label"
  >
    Error
  </span>
</div>
`;
